File: week02/example/ltc46_permutations.py
from typing import List


class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        used = [False] * n
        ans = []
        tmp_permutation = []

        def recur(nums, pos):
            if pos == n:
                ans.append(tmp_permutation.copy())
                return
            for i in range(n):
                if not used[i]:
                    tmp_permutation.append(nums[i])
                    used[i] = True
                    recur(nums, pos + 1)
                    used[i] = False
                    tmp_permutation.pop()

        recur(nums, 0)
        return ans


if __name__ == '__main__':
    nums = [1, 2, 3]
    expect_out = [[1, 2, 3], [1, 3, 2], [2, 1, 3], [2, 3, 1], [3, 1, 2], [3, 2, 1]]
    print(Solution().permute(nums))
    # assert Solution().permute(nums) == expect_out

File: week02/example/ltc46_permutations.cpp
//
// Created by BingweiChen on 2022/5/3.
//
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    vector <vector<int>> permute(vector<int> &nums) {
        n = nums.size();
        used = vector<bool>(n, false);
        recur(nums, 0);
        return ans;


    }

private:
    int n;
    vector<int> a;
    vector<bool> used;
    vector <vector<int>> ans;

    void recur(vector<int> &nums, int pos) {
        if (pos == n) {
            ans.push_back(a);
            return;
        }

        for (int i = 0; i < n; i++) {
            if (!used[i]) {
                a.push_back(nums[i]);
                used[i] = true;
                recur(nums, pos + 1);
                used[i] = false;
                a.pop_back();
            }
        }
    }
};